src/modules/schnorr_adaptor/main_impl.h

```diff
@@ -0,0 +1,324 @@
+    secp256k1_sha256_initialize(sha);
+    sha->s[0] = 0x50685e98ul;
+    sha->s[1] = 0x6313905eul;
+    sha->s[2] = 0x6db24fa0ul;
+    sha->s[3] = 0xc8b15c48ul;
+    sha->s[4] = 0x6b318921ul;
+    sha->s[5] = 0x441d8ff3ul;
+    sha->s[6] = 0xa7033a66ul;
+    sha->s[7] = 0xc3545cddul;
+
+    sha->bytes = 64;
+}
+
+/* algo argument for adaptor_nonce_function_bip340 to derive the nonce of Schnorr adaptor signature
+ * by using the correct tagged hash function. */
+static const unsigned char adaptor_bip340_algo[20] = "SchnorrAdaptor/nonce";
+
+static int adaptor_nonce_function_bip340(unsigned char *nonce32, const unsigned char *msg32, const unsigned char *key32, const unsigned char *adaptor, const unsigned char *xonly_pk32, const unsigned char *algo, size_t algolen, void *data) {
+    secp256k1_sha256 sha;
+    unsigned char masked_key[32];
+    int i;
+
+    if (algo == NULL) {
+        return 0;
+    }
+
+    if (data != NULL) {
+        secp256k1_adaptor_nonce_function_bip340_sha256_tagged_aux(&sha);
+        secp256k1_sha256_write(&sha, data, 32);
+        secp256k1_sha256_finalize(&sha, masked_key);
+        for (i = 0; i < 32; i++) {
+            masked_key[i] ^= key32[i];
+        }
+    } else {
+        /* Precomputed TaggedHash("SchnorrAdaptor/aux", 0x0000...00); */
+        static const unsigned char ZERO_MASK[32] = {
+              65, 206, 231, 5, 44, 99, 30, 162,
+              119, 101, 143, 108, 176, 134, 217, 23,
+              54, 150, 157, 221, 198, 161, 164, 85,
+              235, 82, 28, 56, 164, 220, 113, 53
+        };
+        for (i = 0; i < 32; i++) {
+            masked_key[i] = key32[i] ^ ZERO_MASK[i];
+        }
+    }
+
+    /* Tag the hash with algo which is important to avoid nonce reuse across
+     * algorithms. If this nonce function is used in BIP-340 signing as defined
+     * in the spec, an optimized tagging implementation is used. */
+    if (algolen == sizeof(adaptor_bip340_algo)
+            && secp256k1_memcmp_var(algo, adaptor_bip340_algo, algolen) == 0) {
+        secp256k1_adaptor_nonce_function_bip340_sha256_tagged(&sha);
+    } else {
+        secp256k1_sha256_initialize_tagged(&sha, algo, algolen);
+    }
+
+    /* Hash masked-key||pk||msg using the tagged hash as per the spec */
+    secp256k1_sha256_write(&sha, masked_key, 32);
+    secp256k1_sha256_write(&sha, adaptor, 33);
+    secp256k1_sha256_write(&sha, xonly_pk32, 32);
+    secp256k1_sha256_write(&sha, msg32, 32);
+    secp256k1_sha256_finalize(&sha, nonce32);
+    return 1;
+}
+
+const secp256k1_adaptor_nonce_function_hardened secp256k1_adaptor_nonce_function_bip340 = adaptor_nonce_function_bip340;
+
+static int secp256k1_schnorr_adaptor_presign_internal(const secp256k1_context *ctx, unsigned char *presig65, const unsigned char *msg32, const secp256k1_keypair *keypair, secp256k1_adaptor_nonce_function_hardened noncefp, const unsigned char *adaptor, void *ndata) {
```

Comment:
```suggestion
static int secp256k1_schnorr_adaptor_presign_internal(const secp256k1_context *ctx, unsigned char *presig65, const unsigned char *msg32, const secp256k1_keypair *keypair, secp256k1_adaptor_nonce_function_hardened noncefp, const secp256k1_pubkey *adaptor, void *ndata) {
```
The current design allows the user to input an invalid adaptor point. By making this as `secp256k1_pubkey`, we can catch an invalid 33-byte adaptor point during `secp256k1_ec_pubkey_parse`. `secp256k1_ecdsa_adaptor_encrypt` API does the same.

Reply:
I am unsure if the change is really necessary since we parse the `adaptor` point using `secp256k1_ec_pubkey_parse` inside the presign function anyway.

Reply:
The essence of this modification is to make a design choice about serialization/deserialization.

- Option 1: user handles the ser/de-ser of the adaptor point
    - we receive the adaptor point as `secp256k1_pubkey` (in presign & extract_adaptor)
- Option 2: API handles the ser/de-ser of the adaptor point
    - we receive the adaptor point as `unsigned_char *` (in presign & extract_adaptor)

I think it makes sense to leave the ser/de-ser to the user since we already do this for the pubkey point. What’s the libsecp norm here?

Reply:
Just realized there is already a comment regarding this: https://github.com/BlockstreamResearch/secp256k1-zkp/pull/268#discussion_r1442086962

Reply:
Yes, I'd say the libsecp norm is to allow avoiding unnecessary costly recomputations such as pubkey deserialization.